src/libraries/System.Text.Encodings.Web/src/System/Text/Encodings/Web/DefaultHtmlEncoder.cs

```diff
@@ -0,0 +1,211 @@
+
+        private readonly OptimizedInboxTextEncoder _innerEncoder;
+
+        internal DefaultHtmlEncoder(TextEncoderSettings settings)
+        {
+            if (settings is null)
+            {
+                throw new ArgumentNullException(nameof(settings));
+            }
+
+            _innerEncoder = new OptimizedInboxTextEncoder(EscaperImplementation.Singleton, settings.GetAllowedCodePointsBitmap());
+        }
+
+        public override int MaxOutputCharactersPerInputCharacter => 8; // "&#xFFFF;" is worst case for single char ("&#x10FFFF;" [10 chars] worst case for arbitrary scalar value)
+
+        /*
+         * These overrides should be copied to all other subclasses that are backed
+         * by the fast inbox escaping mechanism.
+         */
+
+#pragma warning disable CS0618 // some of the adapters are intentionally marked [Obsolete]
+        private protected override OperationStatus EncodeCore(ReadOnlySpan<char> source, Span<char> destination, out int charsConsumed, out int charsWritten, bool isFinalBlock)
+            => _innerEncoder.Encode(source, destination, out charsConsumed, out charsWritten, isFinalBlock);
+
+        private protected override OperationStatus EncodeUtf8Core(ReadOnlySpan<byte> utf8Source, Span<byte> utf8Destination, out int bytesConsumed, out int bytesWritten, bool isFinalBlock)
+            => _innerEncoder.EncodeUtf8(utf8Source, utf8Destination, out bytesConsumed, out bytesWritten, isFinalBlock);
+
+        private protected override int FindFirstCharacterToEncode(ReadOnlySpan<char> text)
+            => _innerEncoder.GetIndexOfFirstCharToEncode(text);
+
+        public override unsafe int FindFirstCharacterToEncode(char* text, int textLength)
+            => _innerEncoder.FindFirstCharacterToEncode(text, textLength);
+
+        public override int FindFirstCharacterToEncodeUtf8(ReadOnlySpan<byte> utf8Text)
+            => _innerEncoder.GetIndexOfFirstByteToEncode(utf8Text);
+
+        public override unsafe bool TryEncodeUnicodeScalar(int unicodeScalar, char* buffer, int bufferLength, out int numberOfCharactersWritten)
+            => _innerEncoder.TryEncodeUnicodeScalar(unicodeScalar, buffer, bufferLength, out numberOfCharactersWritten);
+
+        public override bool WillEncode(int unicodeScalar)
+            => !_innerEncoder.IsScalarValueAllowed(new Rune(unicodeScalar));
+#pragma warning restore CS0618
+
+        /*
+         * End overrides section.
+         */
+
+        private sealed class EscaperImplementation : ScalarEscaperBase
+        {
+            internal static readonly EscaperImplementation Singleton = new EscaperImplementation();
+
+            private EscaperImplementation() { }
+
+            internal override int EncodeUtf8(Rune value, Span<byte> destination)
+            {
+                if (value.Value == '<')
+                {
+                    if (!SpanUtility.TryWriteBytes(destination, (byte)'&', (byte)'l', (byte)'t', (byte)';')) { goto OutOfSpace; }
```

Comment:
Maybe I'm missing something crucial here, but wouldn't it be better to pass an uint directly and do  BinaryPrimitives.ReverseEndianness if required? ReverseEndianess is a bswap, compared to multiple shift and or in TryWriteBytes or even pass the appropriate reversed endianess version directly if required?
If I read the sharplab asm correcctly, apparently having something like
```csharp
uint v = (byte)'&' << 24 | (byte)'l' << 16 | (byte)'t' << 8 | (byte)';';
if (BitConverter.IsLittleEndian)
{
     v = BinaryPrimitives.ReverseEndianness(v);
}
```
is by now evaluated as a runtime constant (never realized that before)
https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIAYACY8gOgCEBXAMy5il3YBLAHbYoATwDcNek1YAlDsIyD8MFgGEI+AA6CANnwDKfAG6CwMXNNqNmLRctXqAksr4QdJqOcvWaMgDMTKQMGgwA3jQMMUzBxCgMALIAFACUkdGx2dlGOtjCADzA4hgwAHwMwPocwAwAvAy4GNhgANbY+voQYFWlMADapHQAujY5EwwcIhgMpg0MKSVlaQDkAGSrDIWFDKSJhIvLMGv6WzsM5EgMh0v9axjnuwAcN0f3q5KfWZOxglyLNiCDBaYSmPhlKAsFy4AAywIwhgAosIACaCAppH6/TLUHGTeaNIGiCQABSgqmBgnBAnkMHB/BgKPRBWEVlwKVMaXG+IYAF8AnjeQAVCQAdQpZTY/Q51VqaDm3OxAuo2OyAySMAwAAsIKiXLp9ClNTq9QadPoAPI6FQQYQCACCAHMnbBcLhqTA3PoRCInWkRmrYsR4lcqhAIPoGKLxBLgTBpWUOXkCsV+pVcPlhArpso5p0OCcgzEokL8f9FpmCixYTBhE6dQxyo0UFiyzjS7ycgBVe3YXgsONlXudQROtmowq5jDlFKwAGa/DQcRJMS4bWdFgAcS1dN4sGElhSVeEaQVpgLJx5XZixAA7AwMFBC9ecSquzB9LgYMWcp2b7eD5cJ036vr876TCqfJAA

that would allow you to keep the bytes visible and not require some "strange" magic value.

Reply:
The goal was to keep the call site readable: "I'm writing these bytes in this order." If we want to change the implementation to call `ReverseEndianness` as an implementation detail I'm fine with that. You're right in that the `ReverseEndianness` API was intended to be optimized by the JIT as const input -\> const output.